dbms/src/Storages/Transaction/RegionTable.h

```diff
@@ -164,17 +186,33 @@ class RegionTable : private boost::noncopyable
                                                                     Poco::Logger * log);
 
     /// extend range for possible InternalRegion or add one.
-    void extendRegionRange(const RegionID region_id, const RegionRangeKeys & region_range_keys);
+    void extendRegionRange(RegionID region_id, const RegionRangeKeys & region_range_keys);
+
+
+    void updateSafeTS(UInt64 region_id, UInt64 leader_safe_ts, UInt64 self_safe_ts);
+
+    // unit: ms. If safe_ts diff is larger than 2min, we think the data synchronization progress is far behind the leader.
+    static const UInt64 SafeTsDiffThreshold = 2 * 60 * 1000;
+    bool isSafeTSLag(UInt64 region_id, UInt64 * leader_safe_ts, UInt64 * self_safe_ts);
+    bool canPrintLog()
+    {
+        if (Clock::now() > last_print_log_time + PRINT_LOG_INTERVAL)
+        {
+            last_print_log_time = Clock::now();
+            return true;
+        }
+        return false;
+    }
```

Comment:
make `last_print_log_time` to be a static var in `HandleHttpRequestSyncStatus` instead of here?